### ISSUE TITLE
btoa should accept non-utf8 data

### ISSUE THREAD

**Issue by pierreprinetti** [rustwasm/wasm-bindgen#3632]
### Describe the Bug

**TL;DR: In its current implementation, I understand that [`window.btoa`](https://docs.rs/web-sys/latest/web_sys/struct.Window.html#method.btoa) can't be used to base64-encode binary data.**

To mimic its Web API counterpart, web_sys' `btoa` accepts a `&str` as its argument. However, [as MDN mentions](https://developer.mozilla.org/en-US/docs/Glossary/Base64#javascript_support):

>Base64 is a binary encoding rather than a text encoding, but btoa and atob were added to the web platform before it supported binary data types. As a result, the two functions use strings to represent binary data, with the code point of each character representing the value of each byte

The Web API function [btoa actually accepts something MDN calls "a binary string"](https://developer.mozilla.org/en-US/docs/Web/API/btoa), where each character in the string is treated as a byte of binary data.

**Would it be possible to have an alternative function in `Window` that accepts arbitrary binary data, without forcing the consumer to construct a Rust utf-8 string?**

For the sake of the example, I am going to suggest `window.btoa_from_slice` that accepts a `&[u8]`. Note that I am a Rust newbie and this certainly looks wrong, but you get the idea:

```
pub fn btoa_from_slice(&self, btoa: &[u8]) -> Result<String, JsValue>

pub fn atob_to_slice(&self, atob: &str) -> Result<&[u8], JsValue>
```

Thank you.

**Comment by daxpedda** [comment]
I'm no Unicode expert, but my understanding is that UTF-16 0x00 - 0xFF is valid, but UTF-8 is only valid until 0x7F, 0x80 would describe a two byte character. In addition `DOMString` isn't required to be valid UTF-16, I'm not sure if this play a role here though.
This is why `DOMString` works but `String`/`str` doesn't.

I don't know how to solve this generally in `wasm-bindgen`, maybe introduce a new `DOMString` type and offer an easy conversion from `String` to it. But this requires it's own discussion.

I'm not particularly against an addition like the one you proposed, as long as this is the exception and not the rule. @Liamolucko what do you think about it?

In the meantime, if you are not aware, you can just add it yourself when you need it:
```rust
#[wasm_bindgen]
extern "C" {
    type WindowExt;

    #[wasm_bindgen(catch, method)]
    fn btoa_from_slice(this: &WindowExt, btoa: &[u8]) -> Result<String, JsValue>;
    #[wasm_bindgen(catch, method)]
    fn atob_to_slice(this: &WindowExt, atob: &str) -> Result<&[u8], JsValue>;
}

let window: WindowExt = web_sys::window().unwrap().unchecked_into();
let _ = window.btoa_from_slice(&[128]);

**Comment by pierreprinetti** [comment]
I wasn't aware indeed! Thank you very much.

**Comment by Liamolucko** [comment]
> I don't know how to solve this generally in `wasm-bindgen`, maybe introduce a new `DOMString` type and offer an easy conversion from `String` to it. But this requires its own discussion.

That's basically what [`JsString`](https://docs.rs/js-sys/latest/js_sys/struct.JsString.html) is.

> I'm not particularly against an addition like the one you proposed, as long as this is the exception and not the rule. @Liamolucko what do you think about it?

I'd be fine with it, although I don't really see the point of binding to JavaScript's base64 implementation rather than just using one written in Rust.

It couldn't just be a direct binding though because `atob`/`btoa` don't accept `Uint8Array`s, so it'd have to be a manually-written wrapper around versions that use `JsString`.

**Comment by pierreprinetti** [comment]
> I don't really see the point of binding to JavaScript's base64 implementation rather than just using one written in Rust

In general my assumption is that by using Web APIs I reduce my WASM blob size and I leverage potentially faster native implementations. Am I wrong?

**Comment by daxpedda** [comment]
> It couldn't just be a direct binding though because `atob`/`btoa` don't accept `Uint8Array`s, so it'd have to be a manually-written wrapper around versions that use `JsString`.

Right, so the thing I wrote up there is nonsense really, but replacing `&[u8]` with `JsString` would work. Though I wouldn't be too happy about adding that to `web-sys`.

> > I don't really see the point of binding to JavaScript's base64 implementation rather than just using one written in Rust
> 
> In general my assumption is that by using Web APIs I reduce my WASM blob size and I leverage potentially faster native implementations. Am I wrong?

You are not, but maybe you want to measure that first.
@Liamolucko is right, you would need to use `JsString` for that, which would require you to convert whatever you are having into `[u16]` first (I found some other interesting solutions online with e.g. `FileReader`). Depending on your requirements, this might be too much overhead compared to a pure Rust solution.
